### PR TITLE
Fix ProcessGroupGloo allgather for tensors with shared storage

### LINKED ISSUE #20421: The result of  gloo all_gather error
I use gloo for Model Parallelism, when I use all_gather, the result is error.

There are two process, I expect the all_gather result is [Tensor1, Tensor2], but the result is actually [Tensor1, Tensor1].
The Tensor2 is like this
image
The result  is like this, The senod Tensor in the result should be equal as the Tensor2
image
But before all_gather, I use torch.reshape(torch.tensor(range(xxxx), dtype=torch.float32), [16, 16, 16]) to create two Tensor and all_gather. The result is correctly.
The code is:

```python
for _ in range(stage.get_devices_num()):
        gather_tensor.append(torch.zeros_like(in_slice))
dist.all_gather(gather_tensor, in_slice.contiguous(), group=group)
```

Environment:
     macos
     pytorch                   1.0.1
     pytorch-cpu               1.1.0
     numpy                     1.16.2

PS: We use torch.chunk to split the Tensor and the dim is 0, and all_gather the chunked tensor by gloo, the all_gather  result is error.I think  although I use contiguous() to make memory contiguous, but the it is not effective after I chunk the tensor at dim 0.

### PR DESCRIPTION
**Pull request by mrshenli**
Fix #20421

`ProcessGroupGloo` only requires input/output tensors to be contiguous. Contiguous tensors might not start from the beginning of the underlying storage, e.g., `chunk(..., dim=0)[1]`. The current implementation passes `tensor.storage().data()` ptr to gloo buffer. This leads to wrong results if the tensor has a non-zero storage offset. 

The proposed solution is to use `tensor.data_ptr()` instead. Let's see if this breaks any tests.

cc @qijianan777

### CODE REVIEW

**Comment by mrshenli** [comment]
This might fix #21480 as well, let me add a test.

**Comment by pietern** [comment]
I suppose this is not specific to tensors with shared storage but to all views more generally, right?

**Comment by mrshenli** [comment]
> I suppose this is not specific to tensors with shared storage but to all views more generally, right?

I think it applies to all cases where `tensor.data() != tensor.storage().data()`. So, not necessarily all views, we could still have a view representing the first row of a 2D contiguous tensor not hitting the error.

**Review by pietern — APPROVED** [review]
Can we remove `getDataPointer` in this PR or is it still used somewhere?

**Comment by mrshenli** [comment]
> Can we remove getDataPointer in this PR or is it still used somewhere?

@pietern  I found that `getDataPointer` and `getDataPointers` are only used in `ProcessGroupGloo` for now. So, I moved the change and comments to `getDataPointer` instead.
